Loadtest users GA

kaapstorm opened the pull request:
## Product Description

Makes the "Loadtest users" feature Generally Available for Pro Plan subscriptions and higher.

This change also adds a guard rail to prevent loadtest users from syncing payloads over 50,000 cases.

## Technical Summary

I intend to change the UI to move the "Loadtest factor" input field out of the mobile worker's Basic Info form, and under the "Actions" tab, where it is less likely to cause confusion for users just trying to create a normal mobile worker. That change will follow shortly. I will add the commits to this PR.

:blowfish: :fish: :tropical_fish: 


## Feature Flag

Moves the "Loadtest users" feature flag into GA for Pro Plan subscriptions and higher.

## Safety Assurance

### Safety story

### Automated test coverage

Changes covered by tests.

### QA Plan

This change will go through QA.


### Migrations

- [x] The migrations in this code can be safely applied first independently of the code

**The migration in this PR does not have a reverse action, and cannot be rolled back.**


### Rollback instructions

- [ ] This PR can be reverted after deploy with no further considerations

**The migration in this PR does not have a reverse action, and cannot be rolled back.**


### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

kaapstorm committed:
Remove duplicate imports

kaapstorm committed:
loadtest_users_enabled() checks in one place

kaapstorm committed:
Add LOADTEST_USERS privilege

kaapstorm committed:
Use privilege instead of toggle

kaapstorm committed:
isort

kaapstorm committed:
Add guard rails

kaapstorm committed:
reset_loadtest_factor()

kaapstorm committed:
Move "Loadtest factor" to "Actions" tab

kaapstorm committed:
Drop ENABLE_LOADTEST_USERS toggle

kaapstorm committed:
Drop some type hints

proteusvacuum commented on corehq/ex-submodules/casexml/apps/phone/data_providers/case/livequery.py:
```diff
@@ -113,6 +115,11 @@ def do_livequery(timing_context, restore_state, response, async_task=None):
+    # is created.
+    if loadtest_users_enabled(restore_state.domain):
+        for owner_id in owner_ids:
+            reset_loadtest_factor(owner_id)
```
This is a code path that is hit very often, do you think that pulling every owner synchronously could pose a performance problem? Otherwise I'd advocate to just kick off a task or something that does this with the list, or do the update in bulk.

proteusvacuum commented on corehq/toggles/__init__.py:
```diff
@@ -1055,14 +1055,6 @@ def _enable_search_index(domain, enabled):
     namespaces=[NAMESPACE_DOMAIN]
 )
 
-ENABLE_LOADTEST_USERS = StaticToggle(
```
🎉

kaapstorm committed:
If you're going to use type hints ...

... then include the return value

kaapstorm committed:
Call `reset_loadtest_factor()` asynchronously

kaapstorm committed:
Merge branch 'master' into nh/loadtest_users_ga

ctsims commented:
Question: Is 50k cases too low for Dimagi's team? 

I haven't used this feature in a bit, but that doesn't actually sound like an "extreme" amount of cases from my perspective. For external users I think it's a fine limit, but I believe internally I've needed to create more than that with some frequency.

Also: Can y'all share the documentation we are going to share with external users about this feature? Its behavior is not entirely intuitive. 

Along with the last point, I'd like to understand the safety story a bit from the user's perspective, as this feature has some rough edges. Three Examples: 

1. Previous, this feature had some limitations around parent/child cases, but I think that it did get updated at some point such that it creates relationships correctly. That does illustrate that there could be other potential configurations where a "load test factor" restore could be invalid. Have we validated that there aren't any other configurations where the load test won't work as expected?
1. The phone doesn't exactly work "normally" after a loadtest user is restored. Every other sync is a 412, for instance, which makes sense, but as a user it's not quite clear what to expect
1. This feature creates a kind of "virtual" case which doesn't exist, but AFAIK the phone currently doesn't have anything preventing a form from being submitted for that case, which will make a new, likely malformed case with only the "delta" of case properties, potentially breaking their sync. Is there any safety around that which has been established?

kaapstorm committed:
Merge branch 'master' into nh/loadtest_users_ga

kaapstorm committed:
Merge branch 'master' into nh/loadtest_users_ga

kaapstorm committed:
Allow a lot more cases

kaapstorm committed:
Tear down user to avoid resource conflicts

kaapstorm commented:
Thanks @ctsims, those are great questions.

* I have bumped the max case limit from 50,000 to 500,000. 
* The documentation is on [internal Confluence here](https://confluence.dimagi.com/display/saas/Loadtest+Users) (for now).
* This is due to go through QA on Staging. I will try to recruit their help to evaluate the stafety story from the user's perspective, and then see whether we can either address some of the issues with code, or with documentation, or both.

ctsims commented:
Thanks for providing that context, @kaapstorm .

Can you create a doc outlining the Safety Story for this? I have more input but they're not code-review related so I don't think this is the right place to discuss. I think having a good Safety Story doc it's an important part of the FF -> GA transition, and while QA will probably be helpful, I think it's an antipattern for FF's especially to hope QA will find issues. Feature Flags are full of limitations and implementation complexities that are invisible to QA, and I think that developers have a higher onus to identify sources of risk proactively. QA should ideally be an *answer* to individual points of risk in a safety story, not a way to surface questions.

For all of these areas under discussion in the safety story I think whoever is looking through from SaaS's side for approval should be trying to have these risks actively surfaced and communicated, along with the details. 

IE: for the case load limit, how was the number chosen? For forms submitted against duplicates, what's an example of the behavior so someone can identify the risk? 

It's fine if the answers for these things aren't highly specific or detailed, ("50k cases worked when I tried it and seems like enough" is a fine and honest starting point to communicate) but it's really helpful for a reader to understand where they came from to see if they want to ask someone to dig in further.

proteusvacuum commented:
I hadn't realized that we accepted form submissions against the loadtest cases, so I agree with the point Clayton is making here. I think we should proactively come up with some mitigation strategies for this probably in a different doc somewhere.

ctsims commented:
Thanks @kaapstorm. I think this can be a good template for how we approach assessing and communicating about risk and safety since it's a relatively small feature with a strong value prop.

kaapstorm committed:
Fix "Update Information" button

`StrictButton` with `type="submit"` avoids a silent failure in
hq.helpers.js that prevents the form from being submitted.

biyeun commented on corehq/privileges.py:
```diff
@@ -85,6 +85,8 @@
 
 LITE_RELEASE_MANAGEMENT = 'lite_release_management'
 
+LOADTEST_USERS = 'loadtest_users'
```
hey @kaapstorm! is it possible to PR this migration of adding a new privilege separately so that it can quickly get merged and unblock any merge conflicts happening with migrations on staging? I think on SaaS we generally have a policy of putting migrations in PRs separate from code changes to avoid issues like that and also ensure that the code changes can be easily reverted if needed. Thank you! ^_^

kaapstorm replied:
That is an excellent idea! Thank you, will do.

kaapstorm commented:
Based on @biyeun 's [suggestion](https://github.com/dimagi/commcare-hq/pull/31095#discussion_r833432186) I'm closing this, and opening one PR for the migration, and a second for the rest of the commits.
